packages/js/product-editor/src/hooks/use-product-template/use-product-template.ts

```diff
@@ -16,13 +16,48 @@ declare global {
 	}
 }
 
+const matchesAllTemplateMetaFields = (
+	templateMeta: Metadata< string >[],
+	productMeta: Metadata< string >[]
+) =>
+	templateMeta.every( ( item ) =>
+		productMeta.find(
+			( productMetaEntry ) =>
+				productMetaEntry.key === item.key &&
+				productMetaEntry.value === item.value
+		)
+	);
+
+function templateDataMatchesProductData(
+	productTemplate: ProductTemplate,
+	product: Partial< Product >
+): unknown {
+	return Object.entries( productTemplate.productData ).every(
+		( [ key, value ] ) => {
+			if ( key === 'meta_data' ) {
+				return matchesAllTemplateMetaFields(
+					value,
+					product.meta_data || []
+				);
+			}
+
+			return product[ key ] === value;
+		}
+	);
+}
+
 export const useProductTemplate = (
 	productTemplateId: string | undefined,
-	productType: ProductType | undefined
+	product: Partial< Product > | undefined
 ) => {
+	if ( ! product ) {
+		return { productTemplate: undefined, isResolving: true };
+	}
```

Comment:
Wouldn't we return a matching product template if one was found that matched `productTemplateId`, even if no product was passed in?

And, wouldn't `isResolving` be `false` in that case?

Or, is this to handle the scenario where the product itself hasn't been resolved yet? I wonder if we can/should distinguish between those two cases? For example, we could pass in `undefined` for `product` if the product was still resolving and `null` if the product is resolved and does not exist.